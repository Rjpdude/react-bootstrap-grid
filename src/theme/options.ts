export interface ThemeOptions {
  /**
   * The `margin` and `padding` in pixels for all grid elements.
   *
   * @default 15
   */
  spacing: number
  /**
   * The maximum width in pixels of the small breakpoint container.
   *
   * @default 576
   */
  sm: number
  /**
   * The width in pixels to assign to the small breakpoint container.
   *
   * @default 540
   */
  smWidth: number
  /**
   * The maximum width in pixels of the medium breakpoint container.
   *
   * @default 768
   */
  md: number
  /**
   * The width in pixels to assign to the medium breakpoint container.
   *
   * @default 720
   */
  mdWidth: number
  /**
   * The maximum width in pixels of the large breakpoint container.
   *
   * @default 992
   */
  lg: number
  /**
   * The width in pixels to assign to the large breakpoint container.
   *
   * @default 960
   */
  lgWidth: number
  /**
   * The maximum width in pixels of the extra-large breakpoint container.
   *
   * @default 1200
   */
  xl: number
  /**
   * The width in pixels to assign to the extra-large breakpoint container.
   *
   * @default 1140
   */
  xlWidth: number
}

export function getDefaultTheme(): ThemeOptions {
  return {
    spacing: 15,
    sm: 576,
    smWidth: 540,
    md: 768,
    mdWidth: 720,
    lg: 992,
    lgWidth: 960,
    xl: 1200,
    xlWidth: 1140
  }
}
