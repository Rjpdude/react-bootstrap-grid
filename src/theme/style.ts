import { css, createGlobalStyle } from 'styled-components'

export const Global = css`
  html {
    box-sizing: border-box;
    -ms-overflow-style: scrollbar;
  }

  *,
  *::before,
  *::after {
    box-sizing: inherit;
  }
`

export const GridThemeStyle = createGlobalStyle`
  ${Global}
`
