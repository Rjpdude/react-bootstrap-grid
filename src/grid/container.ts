import { CSSObject } from 'styled-components'
import { configureElement } from '../element/element'
import { ThemeOptions } from '../theme/options'

export interface ContainerProps {
  size: 'fluid' | 'sm' | 'md' | 'lg' | 'xl'
}

export const Container = configureElement<ContainerProps>(
  'Container', ({ ownProps, themeProps }) => {
    const containerStyle = initialContainerStyle(themeProps)
    const size = ownProps.size

    if (size !== 'fluid') {
      const sizeIndex = !size ? -1 : containerSizeBreakpoints.indexOf(size)
      const breakpoints =
        sizeIndex === -1
          ? containerSizeBreakpoints
          : containerSizeBreakpoints.filter((_, index) => index >= sizeIndex)

      breakpoints.forEach((breakpoint) => {
        containerStyle[`@media (min-width: ${themeProps[breakpoint]}px)`] = {
          maxWidth: `${themeProps[`${breakpoint}Width`]}px`
        }
      })
    }

    return containerStyle
  },
  ['size']
)

function initialContainerStyle({ spacing }: ThemeOptions): CSSObject {
  const padding = `${spacing}px`
  return {
    width: '100%',
    paddingRight: padding,
    paddingLeft: padding,
    marginRight: 'auto',
    marginLeft: 'auto'
  }
}

const containerSizeBreakpoints = ['sm', 'md', 'lg', 'xl']
