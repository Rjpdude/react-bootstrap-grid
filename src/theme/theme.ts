import { createElement, createContext, PropsWithChildren } from 'react'
import { ThemeOptions, getDefaultTheme } from './options'
import { GridThemeStyle } from './style'

export const gridThemeContext = createContext(getDefaultTheme())

export function GridTheme(props: PropsWithChildren<Partial<ThemeOptions>>) {
  const style = createElement(GridThemeStyle)
  const context = {
    value: getMergedTheme(props)
  }
  return createElement(
    gridThemeContext.Provider,
    context,
    props.themeStyle !== false && style,
    props.children
  )
}

export function getMergedTheme(themeProps: Partial<ThemeOptions>) {
  const theme = getDefaultTheme()

  Object.keys(theme).forEach((key) => {
    if (themeProps[key]) {
      Object.assign(theme, {
        [key]: themeProps[key]
      })
    }
  })

  return theme
}
