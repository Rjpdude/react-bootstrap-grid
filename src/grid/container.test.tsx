import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import { GridTheme } from '../theme/theme'
import { Container } from './container'

test('Default container', () => {
  const output = renderer
    .create(
      <Container>
        <p>child</p>
      </Container>
    )
    .toJSON()
  expect(output).toMatchSnapshot()
})

test('Fluid container', () => {
  const output = renderer
    .create(
      <Container fluid={true}>
        <p>child</p>
      </Container>
    )
    .toJSON()
  expect(output).toMatchSnapshot()
})

test('Breakpoints', () => {
  const output = renderer.create(<Container />).toJSON()

  // small
  expect(output).toHaveStyleRule('max-width', '540px', {
    media: '(min-width:576px)'
  })
  // medium
  expect(output).toHaveStyleRule('max-width', '720px', {
    media: '(min-width:768px)'
  })
  // large
  expect(output).toHaveStyleRule('max-width', '960px', {
    media: '(min-width:992px)'
  })
  // extra-large
  expect(output).toHaveStyleRule('max-width', '1140px', {
    media: '(min-width:1200px)'
  })
})

test('Breakpoints (fluid)', () => {
  const outputFluid = renderer.create(<Container fluid={true} />).toJSON()

  // small
  expect(outputFluid).not.toHaveStyleRule('max-width', '540px', {
    media: '(min-width:576px)'
  })
  // medium
  expect(outputFluid).not.toHaveStyleRule('max-width', '720px', {
    media: '(min-width:768px)'
  })
  // large
  expect(outputFluid).not.toHaveStyleRule('max-width', '960px', {
    media: '(min-width:992px)'
  })
  // extra-large
  expect(outputFluid).not.toHaveStyleRule('max-width', '1140px', {
    media: '(min-width:1200px)'
  })
})

test('Applies padding and breakpoints from context', () => {
  const output = renderer
    .create(
      <GridTheme
        spacing={6}
        sm={1}
        smWidth={2}
        md={3}
        mdWidth={4}
        lg={5}
        lgWidth={6}
        xl={7}
        xlWidth={8}
      >
        <Container />
      </GridTheme>
    )
    .toJSON()

  expect(output).toHaveStyleRule('padding-left', '6px')
  expect(output).toHaveStyleRule('padding-right', '6px')

  // small
  expect(output).toHaveStyleRule('max-width', '2px', {
    media: '(min-width:1px)'
  })
  // medium
  expect(output).toHaveStyleRule('max-width', '4px', {
    media: '(min-width:3px)'
  })
  // large
  expect(output).toHaveStyleRule('max-width', '6px', {
    media: '(min-width:5px)'
  })
  // extra-large
  expect(output).toHaveStyleRule('max-width', '8px', {
    media: '(min-width:7px)'
  })
})

test('Applies element style properties', () => {
  const output = renderer
    .create(
      <Container
        p={16}
        justifyContent="between"
        alignSelf="center"
        alignContent="stretch"
      >
        <p>child</p>
      </Container>
    )
    .toJSON()

  expect(output).toMatchSnapshot()
  expect(output).toHaveStyleRule('padding', '16px')
  expect(output).toHaveStyleRule('justify-content', 'space-between')
  expect(output).toHaveStyleRule('align-self', 'center')
  expect(output).toHaveStyleRule('align-content', 'stretch')
})
