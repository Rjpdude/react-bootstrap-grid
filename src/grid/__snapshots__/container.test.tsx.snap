// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Applies element style properties 1`] = `
.c0 {
  width: 100%;
  padding-right: 15px;
  padding-left: 15px;
  margin-right: auto;
  margin-left: auto;
  padding: 16px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
}

@media (min-width:576px) {
  .c0 {
    max-width: 540px;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 720px;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: 960px;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: 1140px;
  }
}

<div
  className="c0 twbs-grid-container"
>
  <p>
    child
  </p>
</div>
`;

exports[`Default container 1`] = `
.c0 {
  width: 100%;
  padding-right: 15px;
  padding-left: 15px;
  margin-right: auto;
  margin-left: auto;
}

@media (min-width:576px) {
  .c0 {
    max-width: 540px;
  }
}

@media (min-width:768px) {
  .c0 {
    max-width: 720px;
  }
}

@media (min-width:992px) {
  .c0 {
    max-width: 960px;
  }
}

@media (min-width:1200px) {
  .c0 {
    max-width: 1140px;
  }
}

<div
  className="c0 twbs-grid-container"
>
  <p>
    child
  </p>
</div>
`;

exports[`Fluid container 1`] = `
.c0 {
  width: 100%;
  padding-right: 15px;
  padding-left: 15px;
  margin-right: auto;
  margin-left: auto;
}

<div
  className="c0 twbs-grid-container"
>
  <p>
    child
  </p>
</div>
`;
